crates/timers/src/callback.rs

```diff
@@ -1,8 +1,25 @@
 //! Callback-style timer APIs.
 
+use js_sys::Reflect;
 use wasm_bindgen::prelude::*;
-use wasm_bindgen::JsCast;
-use web_sys::window;
+use wasm_bindgen::{throw_val, JsCast, JsValue};
+use web_sys::{Window, WorkerGlobalScope};
+
+macro_rules! global {
+    ($global:ident, $fun:expr) => {{
+        let global: JsValue = js_sys::global().into();
+
+        if Reflect::has(&global, &String::from("Window").into()).unwrap_throw() {
```

Comment:
I think this would be better written like this:

```rust
match js_sys::global().dyn_into::<Window>() {
    Ok(window) => WindowOrGlobal::Window(window),
    Err(global) => {
        let worker = global.dyn_into::<WorkerGlobalScope>().unwrap_throw();
        WindowOrGlobal::Worker(worker)
    },
}
```

Reply:
This doesn't work, `dyn_into` will use `typeof Window` inside of the generated javascript, which will throw an error saying that `Window` doesn't exist.

Reply:
Hmmm, that's a really tricky problem. Let me think about that.

Reply:
Okay, so I think we should use this solution as a temporary measure until we get better support in wasm-bindgen:

```rust
impl WindowOrWorker {
    fn new() -> Self {
        #[wasm_bindgen]
        extern "C" {
            type Global;

            #[wasm_bindgen(method, getter, js_name = Window)]
            fn window(this: &Global) -> JsValue;

            #[wasm_bindgen(method, getter, js_name = WorkerGlobalScope)]
            fn worker(this: &Global) -> JsValue;
        }

        let global: Global = js_sys::global().unchecked_into();

        if !global.window().is_undefined() {
            Self::Window(global.unchecked_into())

        } else if !global.worker().is_undefined() {
            Self::Worker(global.unchecked_into())

        } else {
            panic!("Only supported in a browser or web worker");
        }
    }
}
```

Reply:
May I ask why? What's wrong with `Reflect`?

Reply:
It's slow, because it's more dynamic, it has to do more checks, and it has to serialize strings to JS. This way is faster.

Reply:
May I ask why? What's wrong with `Reflect`?

Reply:
It's slow, because it's more dynamic, it has to do more checks, and it has to serialize strings to JS. This way is faster.

Reply:
Aha! Thanks for the expanation!

Reply:
The extra checks involve both a `try`/`catch` in JS *and* some extra checks on the Rust side, you can see the performance difference clearly by looking at the generated JS code:

```js
export const __wbg_Window_28cae7b222075c59 = function(arg0) {
    const ret = getObject(arg0).Window;
    return addHeapObject(ret);
};
```

vs

```js
let cachedTextDecoder = new TextDecoder('utf-8');

function getStringFromWasm(ptr, len) {
    return cachedTextDecoder.decode(getUint8Memory().subarray(ptr, ptr + len));
}

function getCachedStringFromWasm(ptr, len) {
    if (ptr === 0) {
        return getObject(len);
    } else {
        return getStringFromWasm(ptr, len);
    }
}

function handleError(e) {
    wasm.__wbindgen_exn_store(addHeapObject(e));
}

export const __wbg_get_003e1b80a63de7c5 = function(arg0, arg1) {
    try {
        const ret = Reflect.get(getObject(arg0), getObject(arg1));
        return addHeapObject(ret);
    } catch (e) {
        handleError(e)
    }
};
```

Reply:
Done, again I just double-checked and copy pasted.
Yeah I was looking at it right now, I find this very strange, but I didn't delve enough into `wasm_bindgen` to really know my way around all of this, anyways, thanks for helping out and for the explanation!

Reply:
As a general rule, if you know the properties ahead of time you should use static bindings with `extern "C" { ... }`.

So you would only use `Reflect` when you need to look up things dynamically, like `Reflect::get(&foo, &bar)`

The way to think of it is that `extern "C" { ... }` corresponds to `foo.bar` in JS, whereas `Reflect` corresponds to `foo[bar]` in JS.

Reply:
I see, thanks!

Reply:
This is good advice and should live in the wasm-bindgen book somewhere I think.